pkg/routes/issuermetadata.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package routes
+
+import (
+	"net/http"
+
+	restful "github.com/emicklei/go-restful"
+
+	"k8s.io/klog"
+)
+
+// This code is in package routes because many controllers import
+// pkg/serviceaccount, but are not allowed by import-boss to depend on
+// go-restful. All logic that deals with keys is kept in pkg/serviceaccount,
+// and only the rendered JSON is passed into this server.
+
+const (
+	// OpenIDConfigPath is the URL path at which the API server serves
+	// an OIDC Provider Configuration Information document, corresponding
+	// to the Kubernetes Service Account key issuer.
+	// https://openid.net/specs/openid-connect-discovery-1_0.html
+	OpenIDConfigPath = "/.well-known/openid-configuration"
+	// JWKSPath is the URL path at which the API server serves a JWKS
+	// containing the public keys that may be used to sign Kubernetes
+	// Service Account keys
+	JWKSPath = "/openid/v1/jwks"
+	// cacheControl is the value of the Cache-Control header. Overrides the
+	// global `private, no-cache` setting.
+	cacheControl = "public, max-age=86400" // 1 day
+)
+
+// IssuerMetadataServer is an HTTP server for metadata of the KSA token issuer.
+type IssuerMetadataServer struct {
+	metadataJSON []byte
+	keysetJSON   []byte
+}
+
+// NewServer creates a new IssuerMetadataServer.
+// The issuer is the OIDC issuer; keys are the keys that may be used to sign
+// KSA tokens.
+func NewIssuerMetadataServer(metadataJSON, keysetJSON []byte) *IssuerMetadataServer {
+	return &IssuerMetadataServer{
+		metadataJSON: metadataJSON,
+		keysetJSON:   keysetJSON,
+	}
+}
+
+// Install adds this server to the request router c.
+func (s *IssuerMetadataServer) Install(c *restful.Container) {
+	// Container.Add "will detect duplicate root paths and exit in that case",
+	// so we need a root for /.well-known/openid-configuration to avoid conflicts.
+	cfg := new(restful.WebService)
+	cfg.Path(OpenIDConfigPath).Route(
+		cfg.GET("").
+			To(fromStandard(s.serveConfiguration)).
+			Doc("get serviceaccount issuer OIDC configuration").
+			Operation("getServiceAccountIssuerMetadata"))
```

Comment:
should we add `Produces(...)` directives that indicate the produced mime-types, so requests enforce Accept headers correctly?

Reply:
should we also add `Returns(...)` directives that indicate successful responses? currently, the openapi doc for these endpoints only indicate 401 as a potential response:
```json
{
  "get": {
    "description": "get serviceaccount issuer OIDC configuration",
    "schemes": [
      "https"
    ],
    "tags": [
      "WellKnown"
    ],
    "operationId": "getServiceAccountIssuerMetadata",
    "responses": {
      "401": {
        "description": "Unauthorized"
      }
    }
  }
}
```

Reply:
Yeah that's a good idea. What codes do we need to add here? Just 200 and 500 since 401 is already included? Is there a common set we can install?

Reply:
Added 200 and 501 for now, and the Produces directives for the mime types.

Reply:
Guessing the 401 is configured as one of the `CommonResponses` that the kube-openapi generator can be configured with

Reply:
Done.